jax/abstract_arrays.py

```diff
@@ -248,3 +255,167 @@ def _make_concrete_python_scalar(x):
   ad_util.jaxval_zeros_likers[t] = _zeros_like_python_scalar
 
 core.literalable_types.update(dtypes.python_scalar_dtypes.keys())
+
+def to_index(x):
+  """Like operator.index, but allowing polymorphic dimensions.
+  Not implemented as `Poly.__index__`, since operator.index only allows ints."""
+  return x if isinstance(x, Poly) else op.index(x)
+
+# TODO remove remaining usages:
+def is_polymorphic(shape):
+  return any(map(lambda d: type(d) is Poly, shape))
```

Comment:
`isindex` is kinda slow, so unless you think inheritance may be present (which IIUC it shouldn't be here), I think it's better to use `type(x) is Poly`.

Reply:
Sure, done!